### ISSUE TITLE
l1L1GtObjectMap misconfigured in workflow 5.2 step 1

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#15977]
The module labelled `l1L1GtObjectMap` is configured as

```
cms.EDProducer("ConvertObjectMapRecord",
    L1GtObjectMapTag = cms.InputTag("hltL1GtObjectMap")
)
```

and for workflow 5.2 (and probably others) only appears on path `L1Reco_step`. However, `hltL1GtObjectMap` only appears on paths `HLTriggerFirstPath`, `HLT_Physics_v1`, `HLT_ZeroBias_v1` and `HLTriggerFinalPath`. The schedule for the job is

```
process.schedule = cms.Schedule(*[ process.generation_step, process.genfiltersummary_step, process.simulation_step, process.reconstruction_befmix_step, process.digitisation_step, process.L1simulation_step, process.digi2raw_step, process.L1Reco_step, process.reconstruction_step, process.eventinterpretaion_step, process.HLTriggerFirstPath, process.HLT_Physics_v1, process.HLT_ZeroBias_v1, process.HLTriggerFinalPath, process.HLTAnalyzerEndpath, process.prevalidation_step, process.validation_step, process.dqmoffline_step, process.FEVTDEBUGHLToutput_step, process.DQMoutput_step ])
```

where we can see that all the HLT paths run after `L1Reco_step` which means the data request for `l1L1GtObjectMap` had never been obtainable.

However, now that we have code that can run Paths in parallel, this cross L1/HLT dependency was caught as causing a circular data/path dependency.

**Comment by Dr15Jones** [comment]
Looking further, is this meant to get the value from the source? If so, then I'm guessing what is really wanted is to specify `cms.InputTag.skipCurrentProcess()`.

**Comment by Dr15Jones** [comment]
assign l1

**Comment by Dr15Jones** [comment]
assign hlt

**Comment by Dr15Jones** [comment]
assign reconstruction

**Comment by Martin-Grunewald** [comment]
Hi
```hltL1GtObjectMap``` is for the Run-1/Legacy L1 trigger, and is run within the Run-1/Legacy HLT
("Fake" and "Fake1" [stage-1] menus).
```hltL1GtObjectMap``` gets ins own input from the unpacked L1 raw data.
It does not (can not) depend on L1Reco.
I would assume L1Reco could be intentionally looking at the stuff unpacked within HLT,
ie, by ```hltL1GtObjectMap``` or alternatively L1Reco should do its own unpacking.

**Comment by Dr15Jones** [comment]
@Martin-Grunewald I think I must not have been clear. `hltL1GtObjectMap` does not depend on L1Reco. However, because of the changes to allow multiple modules to run simultaneously for the same event, we have to make all module dependencies work as if they were run unscheduled. For that case, `l1L1GtObjectMap`, which is on a Path that was previously run before all HLT Paths, now will try to pick up `l1L1GtObjectMap` from the present Process instead of from the Source.

This should have been found earlier, but the code that looks for such improper dependencies was not sufficient.

**Comment by Martin-Grunewald** [comment]
Well, if ```l1GtObjectMap``` looks for an HLT product, namely ```L1GtObjectMapTag = cms.InputTag("hltL1GtObjectMap")```, then it can meaningfully run only AFTER the HLT has run.

So I assume the config is still fine but the dependency check was wrong?

**Comment by Dr15Jones** [comment]
I'm afraid the dependency check is right. About a year ago, in order to support the full design of the threaded framework, we required that all data requests (whether to scheduled or unscheduled) had to have the exact same behavior as if all EDProducers in a job were being run unscheduled. 

Therefore the meaning of asking for a data product without giving a process name was slightly changed. Previously it would mean "search backwards in process order and find the most recent process that has ALREADY put the data product into the event". The new interpretation is "search backward in process order and find the most recent process that WILL put the data product into the event".  The second form means if a module in the present process says it can make the data product then we always wait until that module is run (or all its Paths are skipped) to see if it actually puts the data into the event.

Last year when we changed the meaning we made it so an exception would go off if the schedule had a problem. Unfortunately that checker had a flaw which missed a few cases I'm now discovering.
